fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/ArrayContainToArrayOverlap.java

```diff
@@ -0,0 +1,95 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.expression.rules;
+
+import org.apache.doris.nereids.rules.expression.ExpressionRewriteContext;
+import org.apache.doris.nereids.rules.expression.ExpressionRewriteRule;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Or;
+import org.apache.doris.nereids.trees.expressions.functions.scalar.ArrayContains;
+import org.apache.doris.nereids.trees.expressions.functions.scalar.ArraysOverlap;
+import org.apache.doris.nereids.trees.expressions.literal.ArrayLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.trees.expressions.visitor.DefaultExpressionRewriter;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * array_contains ( c_array, '1' )
+ *  OR array_contains ( c_array, '2' )
+ * =========================================>
+ * array_overlap(c_array, ['1', '2'])
+ */
+public class ArrayContainToArrayOverlap extends DefaultExpressionRewriter<ExpressionRewriteContext> implements
+        ExpressionRewriteRule<ExpressionRewriteContext> {
+
+    public static final ArrayContainToArrayOverlap INSTANCE = new ArrayContainToArrayOverlap();
+
+    private static final int REWRITE_PREDICATE_THRESHOLD = 2;
+
+    @Override
+    public Expression rewrite(Expression expr, ExpressionRewriteContext ctx) {
+        return expr.accept(this, ctx);
+    }
+
+    @Override
+    public Expression visitOr(Or or, ExpressionRewriteContext ctx) {
+        List<Expression> disjuncts = ExpressionUtils.extractDisjunction(or);
+        List<Expression> contains = disjuncts.stream()
+                .filter(this::isValidArrayContains)
+                .collect(Collectors.toList());
+        Map<Expression, Set<Literal>> containLiteralSet = new HashMap<>();
+        contains.forEach(containFunc ->
+                containLiteralSet.computeIfAbsent(containFunc.child(0), k -> new HashSet<>())
+                            .add((Literal) containFunc.child(1)));
+
+        Builder<Expression> newDisjunctsBuilder = new ImmutableList.Builder<>();
+        containLiteralSet.forEach((left, literalSet) -> {
+            if (literalSet.size() > REWRITE_PREDICATE_THRESHOLD) {
+                newDisjunctsBuilder.add(
+                        new ArraysOverlap(left,
+                                new ArrayLiteral(ImmutableList.copyOf(literalSet))));
+            }
+        });
+
+        disjuncts.stream()
+                .filter(e -> !canCovertToArrayOverlap(e, containLiteralSet))
```

Comment:
u could use `Collect(Collector.partition)` when collect `contains` and when build ArraysOverLap to avoid check all disjunct here.